www/blog/2018/05/29/three-steps-backwards/index.md

```diff
@@ -0,0 +1,246 @@
+---
+
+![](images/bug-dominoes.jpg)
+
+If you've done any digging into Who's On First you'll probably know that each _place_
+has a GeoJSON document and all of these GeoJSON documents are fed into Elasticsearch
+to allow you to ask the [Spelunker](https://spelunker.whosonfirst.org/) questions about these places.
+
+For historical reasons, the Spelunker runs on an old version of Elasticsearch, 2.4.6 to be precise and the 2.4.x series of releases stopped being supported in [February of 2018](https://www.elastic.co/support/eol). Which means it's a good time to start looking at not only upgrading the Spelunker to the most recent
+version of Elasticsearch but also to look at what's coming up in future releases to try
+and make the upgrade as future proofed as it can be.
+
+This was going to be a post about how the migration and upgrade of Elasticsearch happened.
+But that hasn't happened quite yet. Instead this has ended up as a post about data type
+consistency and about wrestling with JSON Schema instead. This is a post about the _fun_ that ensued. It's also a post about the _reverse domino effect_, where to take one step forwards
+you end up having to take multiple steps back at first.
+
+### First Step Backwards - "schema-less document indexing and other false statements"
+
+![](images/scrabble-false.jpg)
+
+Elasticsearch is a great tool and my first choice for when I have a data set in JSON that I
+want to poke around in. If I don't have a data set in JSON I convert it _to_ JSON and then
+poke around in it with Elasticsearch. One day I will find a data set that can't be converted to JSON. That's not to say Elasticsearch is perfect. The
+[documentation](https://www.elastic.co/guide/en/elasticsearch/reference/current/dynamic-mapping.html) makes it all sound very easy ...
+
+> One of the most important features of Elasticsearch is that it tries to get out of your way and let you start exploring your data as quickly as possible. To index a document, you don’t have to first create an index, define a mapping type, and define your fields — you can just index a document and the index, type, and fields will spring to life automatically:
+
+Which is _kind_ of true; _if_ you have a relatively small data set and _if_ that data set
+has consistent data types. Who's On First isn't small and, until recently, had some
+data type inconsistencies. This was a problem.
+
+It was a problem because unless you've defined the mappings for all your JSON document's
+properties, the first time Elasticsearch comes across a JSON property
+it doesn't know about, it tries to work out the property's data type automagically and
+then uses that data type from that point on. So `"foo": "bar"` will be _dynamically mapped_
+so a string, `"bar": 1234` will be mapped to an `integer` and so on. If Elasticsearch then encounters another
+document with the same property but a different data type, if you're lucky it will
```

Comment:
Should this be "you're lucky if it will complain" ?

Reply:
It _should_ read ...

"if you're lucky it will complain. If you're unlucky it will do it's best to coerce the new data type into the one it's already seen"